### PR TITLE
8260589: Crash in JfrTraceIdLoadBarrier::load(_jclass*)

### PR DESCRIPTION
**Pull request by D-D-H**
Could I have a review of this backport that fixes a crash problem?

Although there are many conflicts, I think it's necessary to backport it to 11u since the problem is very easy to reproduce.

And I hope JFR's folks could review it.

In addition to the different context of modified files, I think the following items we should review carefully:

1. use MaxJfrEventId + 101 instead of LAST_TYPE_ID + 1 as the klass id of void.class
2. jdk 11 doesn't support jfr streaming, so I removed the call `JfrTraceIdEpoch::set_changed_tag_state()` in `load_primitive`
3. the Class in jdk 11 doesn't have the field 'hidden', so I removed `writer->write<bool>(false);` in `write_primitive`
4. there are many differences in the API of JfrTraceId between 11u and tip

Thanks,
Denghui

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260589](https://bugs.openjdk.java.net/browse/JDK-8260589): Crash in JfrTraceIdLoadBarrier::load(_jclass*)

### Reviewers
 * [Christoph Langer](https://openjdk.java.net/census#clanger) (@RealCLanger - **Reviewer**) ⚠️ Review applies to 959d27a889efd8f5dbef3244dc190b3439dca060
 * [Markus Grönlund](https://openjdk.java.net/census#mgronlun) (@mgronlun - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk11u-dev pull/47/head:pull/47` \
`$ git checkout pull/47`

Update a local copy of the PR: \
`$ git checkout pull/47` \
`$ git pull https://git.openjdk.java.net/jdk11u-dev pull/47/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 47`

View PR using the GUI difftool: \
`$ git pr show -t 47`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk11u-dev/pull/47.diff">https://git.openjdk.java.net/jdk11u-dev/pull/47.diff</a>

</details>

### CODE REVIEW

**Comment by D-D-H** [comment]
testing:
release & slowdebug jtreg jdk/jfr

**Comment by D-D-H** [comment]
@mgronlun @egahlin 
Sorry for messaging you directly.
Could you help review this backport when you have time?
This problem is easy to trigger.

Thanks,
Denghui

**Comment by RealCLanger** [comment]
Hi @D-D-H,
Unfortunately I can't comment so much on the actual changes in JFR coding. However, I ran your patch through our testing and couldn't spot any regressions. So I'm approving it. However, I think it would be good if Markus or Erik (or any other JFR expert) could throw an eye on it.

**Review by RealCLanger — APPROVED** [review]

**Comment by D-D-H** [comment]
> Hi @D-D-H,
> Unfortunately I can't comment so much on the actual changes in JFR coding. However, I ran your patch through our testing and couldn't spot any regressions. So I'm approving it. However, I think it would be good if Markus or Erik (or any other JFR expert) could throw an eye on it.

Thank you.
Yes, I also think this backport should be reviewed properly by JFR folks.

**Comment by egahlin** [comment]
> > Hi @D-D-H,
> > Unfortunately I can't comment so much on the actual changes in JFR coding. However, I ran your patch through our testing and couldn't spot any regressions. So I'm approving it. However, I think it would be good if Markus or Erik (or any other JFR expert) could throw an eye on it.
> 
> Thank you.
> Yes, I also think this backport should be reviewed properly by JFR folks.

I think it will have to wait for a few weeks. People in Sweden like to take their vacation during the summer :-)

File: src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceId.cpp
Comment (mgronlun) [inline src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceId.cpp:203]: 101? Because traceid class_id_counter = MaxJfrEventId + 101, letting the first dynamically assigned id start at 102.

File: src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp
Comment (mgronlun) [inline src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp:306]: 101?

File: src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp
Comment (mgronlun) [inline src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp:992]: JfrTypeSet::clear() sets _clear_artifacts to true already. Maybe remove that call or remove the adjoining _clear_artifacts = true.

**Review by mgronlun — CHANGES_REQUESTED** [review]

File: src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp
Comment (D-D-H) [reply src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp]: fixed.

File: src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp
Comment (D-D-H) [reply src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp]: fixed.

File: src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceId.cpp
Comment (D-D-H) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceId.cpp]: fixed.

**Comment by D-D-H** [comment]
@mgronlun 
Thank you, please help review the latest version.

**Review by mgronlun — APPROVED** [review]
Looks good.

**Comment by RealCLanger** [comment]
> 
> 
> /integrate

@D-D-H you (still) need to flag the JBS item with jdk11u-fix-request and add a comment.

I wil run the updated version through our testing once more and I'm ready to approve & sponsor after JBS bug was flagged. Thanks.

**Comment by D-D-H** [comment]
> > /integrate
> 
> @D-D-H you (still) need to flag the JBS item with jdk11u-fix-request and add a comment.
> 
> I wil run the updated version through our testing once more and then I'm ready to approve & sponsor after JBS bug was flagged. Thanks.

Thanks for reminding me.
Tagged.

**Pull request closed without merging**
